src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+            => new RegexNegativeLookaheadGroupingNode(
+                openParenToken, questionToken, _currentToken,
+                ParseGroupingEmbeddedExpression(_options & ~RegexOptions.RightToLeft), ParseGroupingCloseParen());
+
+        private RegexNonBacktrackingGroupingNode ParseNonBacktrackingGrouping(RegexToken openParenToken, RegexToken questionToken)
+            => new RegexNonBacktrackingGroupingNode(
+                openParenToken, questionToken, _currentToken,
+                ParseGroupingEmbeddedExpression(_options), ParseGroupingCloseParen());
+
+        private RegexGroupingNode ParseOptionsGroupingNode(
+            RegexToken openParenToken, RegexToken questionToken, RegexToken optionsToken)
+        {
+            // Only (?opts:...) or (?opts) are allowed.  After the opts must be a : or )
+            ConsumeCurrentToken(allowTrivia: false);
+            switch (_currentToken.Kind)
+            {
+                case RegexKind.CloseParenToken:
+                    // Allow trivia after the options and the next element in the sequence.
+                    _options = GetNewOptionsFromToken(_options, optionsToken);
+                    return new RegexSimpleOptionsGroupingNode(
+                        openParenToken, questionToken, optionsToken,
+                        ConsumeCurrentToken(allowTrivia: true));
+
+                case RegexKind.ColonToken:
+                    return ParseNestedOptionsGroupingNode(openParenToken, questionToken, optionsToken);
+
+                default:
+                    return new RegexSimpleOptionsGroupingNode(
+                        openParenToken, questionToken, optionsToken,
+                        CreateMissingToken(RegexKind.CloseParenToken).AddDiagnosticIfNone(
+                            new EmbeddedDiagnostic(WorkspacesResources.Unrecognized_grouping_construct, openParenToken.GetSpan())));
+            }
+        }
+
+        private RegexNestedOptionsGroupingNode ParseNestedOptionsGroupingNode(
+            RegexToken openParenToken, RegexToken questionToken, RegexToken optionsToken)
+            => new RegexNestedOptionsGroupingNode(
+                openParenToken, questionToken, optionsToken, _currentToken,
+                ParseGroupingEmbeddedExpression(GetNewOptionsFromToken(_options, optionsToken)), ParseGroupingCloseParen());
+
+        private static bool IsTextChar(RegexToken currentToken, char ch)
+            => currentToken.Kind == RegexKind.TextToken && currentToken.VirtualChars.Length == 1 && currentToken.VirtualChars[0].Char == ch;
+
+        private static RegexOptions GetNewOptionsFromToken(RegexOptions currentOptions, RegexToken optionsToken)
+        {
+            var copy = currentOptions;
+            var on = true;
+            foreach (var ch in optionsToken.VirtualChars)
+            {
+                switch (ch.Char)
+                {
+                    case '-': on = false; break;
```

Comment:
I couldn't find what prevents invalid options like `-+-IM`. What happens in such case?

Reply:
Such a token is not invalid :).  you can try that out today with the .net regex parser.  The answer is: the token is processed a character at a time (as this loop does).  ```-``` switches it to 'off' mode, ```+``` switches to ```on```.  So this would be ```off```, ```on```, ```off```, then it would see ```IM``` and turn those both off.  

I will comment to make it clear this is intentional.